lib/random_assign_utils.rb

```diff
@@ -1,6 +1,111 @@
 # frozen_string_literal: true
 
 class RandomAssignUtils
+  def self.log_error(automation, message)
+    raise("[discourse-automation id=#{automation.id}] #{message}.")
+  end
+
+  def self.log_info(automation, message)
+    Rails.logger.info("[discourse-automation id=#{automation.id}] #{message}.")
+  end
+
+  def self.automation_script!(context, fields, automation)
+    unless SiteSetting.assign_enabled?
+      log_error(automation, "discourse-assign is not enabled")
+    end
+
+    unless topic_id = fields.dig('assigned_topic', 'value')
+      log_error(automation, "`assigned_topic` not provided")
+    end
+
+    unless topic = Topic.find_by(id: topic_id)
+      log_error(automation, "Topic(#{topic_id}) not found")
+    end
+
+    min_hours = fields.dig('minimum_time_between_assignments', 'value').presence
+    if min_hours && TopicCustomField
+        .where(name: 'assigned_to_id', topic_id: topic_id)
+        .where('created_at < ?', min_hours.to_i.hours.ago)
+        .exists?
+      log_info(automation, "Topic(#{topic_id}) has already been assigned recently")
+      return
+    end
+
+    unless group_id = fields.dig('assignees_group', 'value')
+      log_error(automation, "`assignees_group` not provided")
+    end
+
+    unless group = Group.find_by(id: group_id)
+      log_error(automation, "Group(#{group_id}) not found")
+    end
+
+    users_on_holiday = Set.new(
+      User
+        .where(id:
+          UserCustomField
+          .where(name: 'on_holiday', value: 't')
+          .pluck(:user_id)
```

Comment:
I _think_ the `select` here is better as it will only do one SQL query rather than 2?

```suggestion
          .select(:user_id)
```